tempesta_fw/http.c

```diff
@@ -3043,124 +3096,138 @@ do {									\
+		if (hid == TFW_HTTP_HDR_HOST || hid == xff_hid)
+			continue;
+		if ((hid == TFW_HTTP_HDR_CONTENT_TYPE) && h_ct_replace) {
+			r |= tfw_msg_write(&it, &h_ct);
+			continue;
+		}
 
-	/* Add request-line into the HTTP/1.1 request. */
-	BUG_ON(TFW_STR_EMPTY(&req->uri_path));
-	__h2_msg_hdr_val(&ht->tbl[TFW_HTTP_HDR_H2_METHOD], &method);
+		if (TFW_STR_EMPTY(field))
+			continue;
+		TFW_STR_FOR_EACH_DUP(dup, field, dup_end) {
+			TfwStr *chunk, *chunk_end, hval = {};
 
-	WRITE_STR(dst, &method);
-	WRITE_LIT(dst, " ");
-	if (auth) {
-		WRITE_LIT(dst, S_HTTP);
-		WRITE_STR(dst, &req->host);
-	}
-	WRITE_STR(dst, &req->uri_path);
-	WRITE_LIT(dst, " ");
-	WRITE_LIT(dst, S_VERSION11);
-	WRITE_LIT(dst, S_CRLF);
+			if (unlikely(TFW_STR_PLAIN(dup))) {
+				r = -EINVAL;
+				break;
+			}
 
-	/* Add 'host' as the first header in request (RFC 7230 section 5.4). */
-	WRITE_LIT(dst, S_F_HOST);
-	if (auth) {
-		WRITE_STR(dst, &req->host);
-	} else if (host) {
-		BUG_ON(!TFW_STR_EMPTY(&req->host));
-		__h2_msg_hdr_val(&ht->tbl[TFW_HTTP_HDR_HOST], &req->host);
-		WRITE_HDR(dst, &req->host);
+			hval.chunks = dup->chunks;
+			TFW_STR_FOR_EACH_CHUNK(chunk, dup, chunk_end) {
+				if (chunk->flags & TFW_STR_HDR_VALUE)
+					break;
+				hval.nchunks++;
+				hval.len += chunk->len;
+			}
+			r |= tfw_msg_write(&it, &hval);
+			r |= tfw_msg_write(&it, &dlm);
+			hval.chunks += hval.nchunks;
+			hval.nchunks = dup->nchunks - hval.nchunks;
+			hval.len = dup->len - hval.len;
+			r |= tfw_msg_write(&it, &hval);
+
+			r |= tfw_msg_write(&it, &crlf);
+		}
+		if (unlikely(r))
+			goto err;
 	}
-	WRITE_LIT(dst, S_CRLF);
 
-	/*
-	 * We have already added 'host' header, thus, it won't be needed in
-	 * the writing cycle for all headers below.
-	 */
-	TFW_STR_INIT(&ht->tbl[TFW_HTTP_HDR_HOST]);
+	r |= tfw_msg_write(&it, &h_xff);
+	r |= tfw_msg_write(&it, &h_via);
+	/* Finally close headers. */
+	r |= tfw_msg_write(&it, &crlf);
 
-	FOR_EACH_HDR_FIELD_FROM(fld, fld_end, req, TFW_HTTP_HDR_REGULAR) {
-		if (TFW_STR_EMPTY(fld))
-			continue;
-		TFW_STR_FOR_EACH_DUP(hdr, fld, dup_end)
-			WRITE_HDR(dst, hdr);
-	}
+	if (unlikely(r))
+		goto err;
 
-	/*
-	 * Headers, which should be added unconditionally, inserted in
-	 * the end.
-	 */
-	memcpy_fast(__TFW_STR_CH(&h_via, 2)->data, g_vhost->hdr_via,
-		    g_vhost->hdr_via_len);
-	WRITE_STR(dst, &h_via);
-	WRITE_STR(dst, &h_conn);
-	WRITE_LIT(dst, S_CRLF);
+	T_DBG3("%s: req [%p] converted to http1.1\n", __func__, req);
 
-	T_DBG3("%s: req adjusted, it->hb_len=%lu, it->hdrs_len=%lu,"
-	       " it->hdrs_cnt=%u, dst=[%p], pg=[%p]\n", __func__, it->hb_len,
-	       it->hdrs_len, it->hdrs_cnt, dst, (char *)page_address(pg));
+	old_head = req->msg.skb_head;
+	req->msg.skb_head = new_head;
 
-	WARN_ON_ONCE(dst - (char *)page_address(pg) != it->hdrs_len);
+	/* Http chains might add a mark for the message, keep it. */
+	new_head->mark = old_head->mark;
 
-	r = ss_skb_replace_page(&req->msg.skb_head, pg, it->hdrs_len,
-				it->hb_len);
```

Comment:
The `hb_len` member of `TfwMsgParseIter` structure is not used any more, and can be removed.